libcontainer/cgroups/fs2/io.go

```diff
@@ -43,6 +45,16 @@ func setIo(dirPath string, r *configs.Resources) error {
 			}
 		}
 	}
+	for _, wd := range r.BlkioWeightDevice {
+		// per-device weight is supported since kernel v5.4, commit 795fe54c2a8.
+		if err := fscommon.WriteFile(dirPath, bfqWeight, wd.WeightString()); err != nil {
```

Comment:
This is unsafe to do unconditionally because old kernels called `kstrtoull` on the input string unconditionally -- and since `WeightString` is of the form `major:minor weight` we will end up setting the weight to the last `major` value in `BlkioWeightDevice`. We should check whether `bfq.weight_device` exists before writing these entries.

Reply:
We should also be returning an error if the user has explicitly set `BlkioWeightDevice` when the kernel doesn't support it.

Reply:
> This is unsafe to do unconditionally because old kernels called `kstrtoull` on the input string unconditionally -- and since `WeightString` is of the form `major:minor weight` we will end up setting the weight to the last `major` value in `BlkioWeightDevice`.

Not quite -- it chokes on the first non-numeric character which is not a newline, in our case `:`

```console
root@ubuntu1910:/sys/fs/cgroup/init.scope# echo "100:101 222" > io.bfq.weight 
bash: echo: write error: Invalid argument
root@ubuntu1910:/sys/fs/cgroup/init.scope# echo "100:100 222" > io.bfq.weight 
bash: echo: write error: Invalid argument
root@ubuntu1910:/sys/fs/cgroup/init.scope# echo "10:0 20" > io.bfq.weight 
bash: echo: write error: Invalid argument
root@ubuntu1910:/sys/fs/cgroup/init.scope# echo "10" > io.bfq.weight 
root@ubuntu1910:/sys/fs/cgroup/init.scope# echo "10:" > io.bfq.weight 
bash: echo: write error: Invalid argument
root@ubuntu1910:/sys/fs/cgroup/init.scope# uname -a
Linux ubuntu1910 5.3.0-64-generic #58-Ubuntu SMP Fri Jul 10 19:33:51 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
root@ubuntu1910:/sys/fs/cgroup/init.scope# 
```

> We should check whether `bfq.weight_device` exists before writing these entries.

This only exists on cgroup v1, and this is v2 code. We can read `io.bfq.weight` and if it says `default ...` rather than a number, the per-device weights are available. Or, add `default` when writing a default weight, if it's accepted we have per-device weights.

Reply:
Updated to check if per-device weights are supported. Also added a test case.